### PR TITLE
[refactor] Use one errors package for internal errors where possible

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Refactors operator provisioning code to use one errors package in (almost) all instances. This removes the 3-4 identical implementations for a retry/fail/warning error and makes handling consistent.

This PR also reworks provisioning steps to return errors rather than structs with boolean flags for continue, fail, etc.

This is helpful because in most of the operator code that interacts with the cluster, we are required to distinguish between three cases
* An error occurred and workspace start should be failed
* No real error occurred, but we need to stop the current reconcile and queue a new one (e.g. we updated an object)
* No real error occurred, but a warning should be added to the DevWorkspace

### What issues does this PR fix or reference?
N/A, internal refactor
Closes https://github.com/devfile/devworkspace-operator/issues/787

### Is it tested? How?
Should be no functional changes; DevWorkspaces should get same conditions/messages as previously and failures should be reported in same way as before.

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: pkg/dwerrors/errors.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/devfile/devworkspace-operator/pkg/provision/sync"
+)
+
+type RetryError struct {
+	// Err is the underlying error causing the problem. If nil, it is not included in the output of Error()
+	Err error
+	// Message is a user-friendly string explaining why the error occurred
+	Message string
+	// RequeueAfter represents how long we should wait before checking if storage is ready
```
Comment (AObuchow) [inline pkg/dwerrors/errors.go:28]: The documentation for `RequeueAfter` mentions storage, though RetryError is used for other things (e.g. the routing can produce this type of error), so this should probably be changed to be a bit more general

**Review by AObuchow — COMMENTED** [review]
Looks good to me so far, this is a really awesome and well needed refactor :) 
Will look at this a bit more next week with a fresh set of eyes before giving a final approval.

Also, unless I missed something, `WarningError` usage looks like it still needs to be implemented before merging.

> No real error occurred, but a warning should be added to the DevWorkspace (currently... unimplemented -- TODO before merging)

**Comment by amisevsk** [comment]
Added handling for warning errors as well. This PR now closes https://github.com/devfile/devworkspace-operator/issues/787

File: pkg/provision/storage/asyncStorage.go [hunk 2]
```diff
@@ -121,7 +123,7 @@ func (p *AsyncStorageProvisioner) ProvisionStorage(podAdditions *v1alpha1.PodAdd
 	deploy, err := asyncstorage.SyncWorkspaceSyncDeploymentToCluster(workspace, configmap, pvcName, clusterAPI)
 	if err != nil {
 		if errors.Is(err, asyncstorage.NotReadyError) {
-			return &NotReadyError{
+			return &dwerrors.RetryError{
 				Message:      "waiting for async storage server deployment to be ready",
```
Comment (AObuchow) [inline pkg/provision/storage/asyncStorage.go:127]: Extreme nitpick (that isin't worth immediately addressing IMO..): Some error messages [are now capitalized](https://github.com/devfile/devworkspace-operator/pull/1064/files#diff-bfb432d51dee16f88d16a491fd7b8f24fbe2d56bc53d1036fee90e6fa27e0354R88) while others are not. I imagine the new standard is that errors should start with a capital letter.

When this gets merged, I can make an issue regarding this (i.e. that all DW error messages should start with a capital letter).

**Review by AObuchow — APPROVED** [review]
Looks good to me, awesome refactor @amisevsk  -- the codebase is looking much cleaner with this PR :grin: 

I also tested out getting 2 warnings on a workspace by referencing a non-existent external DWOC + using an undefined variable in my devworkspace, and the warnings were shown as expected:

```
$ kubectl get dw -n $NAMESPACE -w
NAME                       DEVWORKSPACE ID             PHASE      INFO
code-latest-2              workspacefac79a56787e43ea   Starting   Waiting for editor to start (2 warnings)
```

```YAML
$ kubectl describe dw code-latest-2 -n $NAMESPACE
Name:         code-latest-2
Namespace:    devworkspace-controller
(...)
Status:
  Conditions:
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               Error applying external DevWorkspace-Operator configuration: could not fetch external DWOC with name external-dwoc-test in namespace default: DevWorkspaceOperatorConfig.controller.devfile.io "external-dwoc-test" not found
    Status:                True
    Type:                  DevWorkspaceWarning
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               Error processing variable replacements: invalid variables in component dev: 'latest'
    Status:                True
    Type:                  DevWorkspaceWarning
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               DevWorkspace is starting
    Status:                True
    Type:                  Started
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               Resolved plugins and parents from DevWorkspace
    Status:                True
    Type:                  DevWorkspaceResolved
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               Storage ready
    Status:                True
    Type:                  StorageReady
    Last Transition Time:  2023-04-11T18:20:08Z
    Message:               Networking ready
    Status:                True
    Type:                  RoutingReady
    Last Transition Time:  2023-04-11T18:20:09Z
    Message:               DevWorkspace serviceaccount ready
    Status:                True
    Type:                  ServiceAccountReady
    Last Transition Time:  2023-04-11T18:20:09Z
    Message:               DevWorkspace secrets ready
    Status:                True
    Type:                  PullSecretsReady
    Last Transition Time:  2023-04-11T18:20:09Z
    Message:               DevWorkspace deployment ready
    Status:                True
    Type:                  DeploymentReady
    Last Transition Time:  2023-04-11T18:20:09Z
    Message:               Waiting for editor to start
    Status:                False
    Type:                  Ready
  Devworkspace Id:         workspacefac79a56787e43ea
  Main URL:                https://workspacefac79a56787e43ea.apps.ci-ln-btd40d2-72292.origin-ci-int-gce.dev.rhcloud.com/che-code/
  Message:                 Waiting for editor to start (2 warnings)
  Phase:                   Starting
Events:                    <none>
```

Editing the devworkspaces to remove the problems removed the warnings as well :+1: 

One thing that came to mind:
A similar refactor could potentially be done for the devworkspace routing, e.g. we could use `dwerrors.WrapSyncError()` [here](https://github.com/devfile/devworkspace-operator/blob/main/controllers/controller/devworkspacerouting/sync_ingresses.go#L59).

I'm up for doing this refactor myself in a future PR so that we can get this PR merged today :)

**Comment by AObuchow** [comment]
@amisevsk the e2e `restricted-access-devworkspace` test seems to be failing, though I haven't looked into why yet.

**Comment by amisevsk** [comment]
I've ran e2e tests locally multiple times and the tests are passing :shrug:

**Pull request merged**